jt_notes.md

```diff
@@ -0,0 +1,127 @@
+
+**For chas**: more generally, I wish we were working with frame *objects* that have accompanying metadata (such as timestamp!). That would help solve the above issue about timings in a tidy manner, and also make it a lot easier to generalise to other timebases (such as camera timebase, rather than computer). Oh, and it would also help deal with gaps in the timebase (missing frames), which I think would be a useful thing to cope with, even if right now there might not be any concrete scenario in which frames would be dropped. I fear I may not win that argument about frame objects, since it would be quite a big structural change. But, do you have any alternative ideas about how best to cope with the fact that we may well want to work in terms of timebases other than the computer's own timebase?
+
+Remark **for chas:** currently when we schedule a trigger everything hangs, waiting for the appropriate time to arrive. This stalls camera and sync analysis. Don't you think we should really figure out a way of doing this asynchronously, without blocking (or come to a definite conclusion that blocking is the only reliable enough way to do it on the Pi).
+
+**For me: where do the stages come in?** cli.py sets up the usb/serial link, but where is it used, by whom, and when? Should be clearer after a refactor. Either we need to conditionally interact with the stages, or we do it on a more distributed basis, in response to events. That's more how I do it in my code, and I like that approach.
+
+determine_reference_period.py
+
+**For me and chas:** here and elsewhere, I'm happy to help refactor into a class myself. This module definitely needs some sort of refactoring (see "WTF"/"Ugh" comments).
+
+**For chas**: padding flag seems to be used fairly inconsistently (would affect the if test in establish_indices), and when False they *still* seem to include extra ref frames, just fewer of them. Weird.... In fact, do you ever *not *use it, or should we just remove the parameter?
+
+**For chas**: **IMPORTANT:** there is a structural issue here where the period-determination code processes one big block of frame buffers all together, once. That is not a good solution when running live - it will be much better to make a new attempt to determine a period after each individual new frame is received, rather than doing a massive chunk of processing in the loop in establish_indices. I really think this needs a refactor. Otherwise, apart from anything else, we will end up acquiring significantly more frames than we need to before we can establish a period. There are also scenarios where this approach will fail, where it could succeed if it was allowed just to keep trying indefinitely, processing a rolling buffer of recent frames. (Minor related observation: once that is changed, and the loop is removed from this function, the "didn't find a period" error will not be "critical" when running live)
```

Comment:
This could be a place for python generators? But I need to understand them a bit better. Either way, a FIFO rolling buffer is definitely a better solution.

Reply:
Already raised issue #10